### PR TITLE
`execution_mode` shouldn't leak into SDK client

### PR DESCRIPTION
**Pull request by patrickkuo**
`execution_mode` is leaking in to SDK via TransactionBuilder, this PR fixes that.

### CODE REVIEW

File: crates/sui-adapter/src/execution_mode.rs [hunk 1]
```diff
@@ -46,7 +46,7 @@ pub trait ExecutionMode {
         result: Self::ExecutionResult,
     );
 }
-
+#[derive(Clone)]
```
Comment (tnowacki) [inline crates/sui-adapter/src/execution_mode.rs:49]: nit, newline here between the `}` and the `#[derive` 
And you should also give it `Copy` if you are going to give it Clone

Comment (tnowacki) [reply crates/sui-adapter/src/execution_mode.rs:49]: And I don't see any cases where you are calling clone, do you need this?

File: crates/sui-transaction-builder/src/lib.rs [hunk 2]
```diff
@@ -53,9 +53,13 @@ pub trait DataReader {
 }
 
 #[derive(Clone)]
-pub struct TransactionBuilder(pub Arc<dyn DataReader + Sync + Send>);
+pub struct TransactionBuilder<Mode: ExecutionMode>(Arc<dyn DataReader + Sync + Send>, Mode);
```
Comment (tnowacki) [inline crates/sui-transaction-builder/src/lib.rs:56]: Thoughts about using phantom data?

So 
```
pub struct TransactionBuilder<Mode: ExecutionMode>(Arc<dyn DataReader + Sync + Send>, std::marker::PhantomData<Mode>);
```

File: crates/sui-adapter/src/execution_mode.rs (hunk 1 above)
Comment (patrickkuo) [reply crates/sui-adapter/src/execution_mode.rs:49]: don't need this anymore with PhantomData

Comment (tnowacki) [reply crates/sui-adapter/src/execution_mode.rs:50]: With phantom data, you shouldn't need the clone right?

Comment (tnowacki) [reply crates/sui-adapter/src/execution_mode.rs:50]: Spent too much time in Move it seems :) This is how phantom works for Move

Comment (tnowacki) [reply crates/sui-adapter/src/execution_mode.rs:50]: Don't much mind it having clone, just find it surprising. Could you also give it `Copy` too? no reason for it not to have it

**Review by tnowacki — APPROVED** [review]

**Review by 666lcz — APPROVED** [review]
Nice refactoring!!

**Pull request merged**